src/explore-education-statistics-frontend/src/services/permalinkService.ts

```diff
@@ -47,22 +50,163 @@ export interface Permalink {
+  ];
+
+  rowGroups.forEach((rowGroup, index) => {
+    let currentIndex = 0;
+    for (let i = 0; i < locationAndFilterGroups.length; i += 1) {
+      if (
+        locationAndFilterGroups[i].find(
+          element => element.value === rowGroup[0].value,
+        )
+      ) {
+        currentIndex = i;
+        break;
+      }
+    }
+
+    mappedRowGroups[index] = rowGroup.map(({ value }) =>
+      locationAndFilterGroups[currentIndex].find(
+        element => element.value === value,
+      ),
+    );
+    locationAndFilterGroups.splice(currentIndex, 1);
+  });
+
+  columnGroups.forEach((columnGroup, index) => {
+    let currentIndex = 0;
+    for (let i = 0; i < locationAndFilterGroups.length; i += 1) {
+      if (
+        locationAndFilterGroups[i].find(
+          element => element.value === columnGroup[0].value,
+        )
+      ) {
+        currentIndex = i;
+        break;
+      }
+    }
+
+    mappedColumnGroups[index] = columnGroup.map(({ value }) =>
+      locationAndFilterGroups[currentIndex].find(
+        element => element.value === value,
+      ),
+    );
+    locationAndFilterGroups.splice(currentIndex, 1);
+  });
+
+  return {
+    columns: mappedColumns,
+    columnGroups: mappedColumnGroups,
+    rows: mappedRows,
+    rowGroups: mappedRowGroups,
+  };
+};
+
+const mapFullTable = (unmappedFullTable: any) => {
+  return {
+    ...unmappedFullTable,
+    subjectMeta: {
+      ...unmappedFullTable.subjectMeta,
+      indicators: unmappedFullTable.subjectMeta.indicators.map(
+        (indicator: { label: string; unit: string; value: string }) =>
+          new Indicator(indicator),
+      ),
+      locations: unmappedFullTable.subjectMeta.locations.map(
+        (location: { label: string; level: string; value: string }) =>
+          new LocationFilter(location, location.level),
+      ),
+      timePeriodRange: unmappedFullTable.subjectMeta.timePeriodRange.map(
+        (timePeriod: { code: string; label: string; year: number }) =>
+          new TimePeriod(timePeriod),
+      ),
+    },
+  };
+};
+
+const mapPermalink = (unmappedPermalink: any) => {
+  const mappedFullTable = mapFullTable(unmappedPermalink.fullTable);
+
+  return {
+    ...unmappedPermalink,
+    fullTable: mappedFullTable,
+    configuration: {
+      tableHeadersConfig: reverseMapTableHeadersConfig(
+        unmappedPermalink.configuration.tableHeadersConfig,
+        mappedFullTable.subjectMeta,
+      ),
+    },
+  };
+};
+
 export default {
+  createTablePermalink(query: PermalinkCreate): Promise<Permalink> {
+    return dataApi.post('/permalink', query);
+  },
   getPermalink(publicationSlug: string): Promise<Permalink> {
     return dataApi.get(`Permalink/${publicationSlug}`);
   },
-  createTablePermalink(
-    query: PermalinkCreate,
-  ): Promise<{
-    id: string;
-    title: string;
-    created: string;
-    url: string;
-  }> {
-    return dataApi.post('/permalink', query);
-  },
+  mapFullTable,
+  mapPermalink,
```

Comment:
Think the mapping code is too high up in the project structure (as it's all specific for table tool). I would probably relocate this into `modules/table-tool/utils/permalinks`

Reply:
We're aiming to standardize the table rendering and the reason for a lot for this work was to make it less specific to table tool so this should be fine.

Ideally its going to want to be in common,